### ISSUE TITLE
Build flag precedence for Beta/Main MakeCode builds (inc. Extensions) 

### ISSUE THREAD

**Issue by JohnVidler** [microsoft/pxt-microbit#5348]
Its currently not clear how we should specify build flags to include as defaults for MakeCode beta (and main), both as defaults for normal builds, and when we include flags as part of extension settings.

If we (MEF) want to specify a set of flags for Makecode to use over and above the normal set found in `microbit-v2-samples` and/or `codal-microbit-v2` should these be specified in this repository (`pxt-microbit`) somwhere - and if so where?

In CODAL-only builds, we enforce a strict ordering whereby we union the config set in the following order:

1. Target config
2. Application config 
3. Environment variables

Furthermore, if we need to modify the build flags to enable or disable features to use a particular extension, how and in what order do the extensions apply the flags?

For example; if we have set `CODAL_EXAMPLE: "1"` in Extension A, and `CODAL_EXAMPLE: "2"` in Extension B, which value wins to become the final build? Does it matter that lexographically 'A' is before 'B' and might be included first? Do we also signal to the user somehow in PXT that the extension may conflict based on the config it uses?

Is the current include order the following?

1. Target config
2. Application config 
3. MakeCode pxt.json
4. Extensions-in-alphanumeric-order pxt.json's
5. Application pxtarget.json
6. Build worker environment variables

**Comment by martinwork** [comment]
It seems 2 extensions, or an extension and project cannot make conflicting definitions.

https://github.com/martinwork/exta/blob/master/pxt.json#L28
https://github.com/martinwork/extb/blob/master/pxt.json#L28
![image](https://github.com/microsoft/pxt-microbit/assets/989126/31d65c17-41a6-4a5c-809d-c5994816e17d)

With matching definitions, exta and extc are OK together.
https://github.com/martinwork/extc/blob/master/pxt.json#L28

Adding the "B" definition to a project that uses exta, at first flashes a temporary red message
`conflict on yotta setting...,`
but eventually
![image](https://github.com/microsoft/pxt-microbit/assets/989126/40adcbe9-5c48-492a-8210-77a685ef167c)
Choosing "Remove this" just displays the same box.

**Comment by JohnVidler** [comment]
Thanks for investigating that @martinwork - did you try overriding a configuration set 'higher up' the list?

Can extensions override stuff we have set in core Makecode, or up in CODAL?

**Comment by martinwork** [comment]
> did you try overriding a configuration set 'higher up' the list?

No, I'm not totally sure I understand what 1, 2, 3  & 6 are. I just thought I could help out by testing what I think I understand!

> Can extensions override stuff we have set in core Makecode, or up in CODAL?

I'm happy to try an example.

**Comment by JohnVidler** [comment]
> No, I'm not totally sure I understand what 1, 2, 3 & 6 are. I just thought I could help out by testing what I think I understand!

1 and 2 are the default configuration json's in CODAL (the target config base and the application codal.json)
3 would be pxtarget.json as part of the Makecode environment build (not clear if config can be set here at all, actually?)

and 6 is the container build instance invoked when using C/C++ extensions, for example

> > Can extensions override stuff we have set in core Makecode, or up in CODAL?

> I'm happy to try an example.

Any of our defaults in https://github.com/lancaster-university/codal-microbit-v2/blob/master/target.json#L9 overidden in an extension would do, like `DEVICE_BLE` and friends.

From what you've investigated here it looks to me like makecode extensions must have configurations that result in unique, non-colliding keys, otherwise we get an error, but we must be able to override the target config somehow, to support features like turning bluetooth on/off for some extensions.

**Comment by martinwork** [comment]
https://makecode.microbit.org/_Vw564TC7LPRv overrides DMESG_SERIAL_DEBUG, CODAL_DEBUG, and DEVICE_DMESG_BUFFER_SIZE in an extension https://github.com/martinwork/extc, to turn on the heap debug messages.

https://makecode.microbit.org/_VvChpm2MKDuT does the same from the project pxt.json.
https://makecode.microbit.org/_UzhKKYeUgf8x just shows DIAGNOSTIC messages.
These non-extension examples may need to be persuaded to rebuild by adding or removing a C++ extension. 

https://makecode.microbit.org/_P5w9zkd915j4 has "DEVICE_BLE": 0. It fails to compile
```
/home/build/prjclone/pxtapp/core/pins.cpp: In function 'void pins::pushButton(DigitalPin)':
/home/build/prjclone/pxtapp/core/pins.cpp:647:13: error: expected type-specifier before 'MicroBitButton'
         new MicroBitButton((PinName)getPin((int)(pin))->name, (int)pin, MICROBIT_BUTTON_ALL_EVENTS, PinMode::PullUp);
             ^~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/MICROBIT.dir/pxtapp/core/pins.cpp.obj] Error 1
```

**Comment by abchatra** [comment]
NO plans to implement.
